From 488a23e01534077f31d08974286766c1da904282 Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Thu, 15 Jun 2023 10:13:46 -0700
Subject: [PATCH] ath12k: Clear the flush flags after ce init  and add few more
      checks for checking flush flags

    Usually when work threads carry out the process of
    reconfiguring the chip which was asserted, a lot of other
    things can happen from upper layers, requesting some
    actions especially when the chip is down. Adding a check
    for chip recovery might help at times, but it need not
    always work because of this worker thread async.

    This change will be checking the chip recovery flag in the mac_op functions.
    If the flag ATH12K_FLAG_CRASH_FLUSH is set then return -ESTUTDOWN.

    This will ensure, when the  restart_work and reset_work threads
    take care of the chip recovery and reset work respectively, they set and
    clear the FLUSH bits at the right places, when the FW is fully down and when
    the FW has been initialized properly and the upper layers such as mac80211
    get failure messages when they send out WMI commands when the chip is
    still down.

    Also, add BUG_ON, in case SSR fails to recover the chip. This is
    because the chip will not be in UP state and it is better to crash
    the system in this case. The goal here is to keep the system in a
    usable state, when an external event disrupts the recovery
    process.

Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 16 +++++++--
 drivers/net/wireless/ath/ath12k/core.h |  1 +
 drivers/net/wireless/ath/ath12k/htc.c  |  2 ++
 drivers/net/wireless/ath/ath12k/mac.c  | 61 ++++++++++++++++++++++++++++++++--
 drivers/net/wireless/ath/ath12k/mhi.c  |  1 +
 drivers/net/wireless/ath/ath12k/pci.c  |  2 ++
 drivers/net/wireless/ath/ath12k/peer.c |  8 +++++
 drivers/net/wireless/ath/ath12k/qmi.c  |  1 +
 drivers/net/wireless/ath/ath12k/wmi.c  |  5 +++
 9 files changed, 93 insertions(+), 4 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -52,6 +52,9 @@ module_param_named(ppe_ds_enable, ath12k
 MODULE_PARM_DESC(ppe_ds_enable, "ppe_ds_enable: 0-disable, 1-enable");
 
 static unsigned int ath12k_recovery_mode = ATH12K_MLO_RECOVERY_MODE0;
+unsigned int ath12k_ssr_failsafe_mode = false;
+module_param_named(ssr_failsafe_mode, ath12k_ssr_failsafe_mode, uint, 0644);
+MODULE_PARM_DESC(ssr_failsafe_mode, "ssr failsafe mode: 0-disable, 1-enable");
 
 bool ath12k_mgmt_rx_reordering = false;
 module_param_named(mgmt_rx_reorder, ath12k_mgmt_rx_reordering, bool, 0644);
@@ -1225,6 +1228,8 @@ static int ath12k_core_start(struct ath1
 		goto err_reo_cleanup;
 	}
 
+	WARN_ON(test_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags));
+	set_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags);
 	/* put hardware to DBS mode */
 	if (ab->hw_params->single_pdev_only) {
 		ret = ath12k_wmi_set_hw_mode(ab, WMI_HOST_HW_MODE_DBS);
@@ -1749,8 +1754,6 @@ static int ath12k_core_reconfigure_on_cr
 	if (ret)
 		return ret;
 
-	clear_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
-
 	ret = ath12k_core_qmi_firmware_ready(ab);
 	if (ret)
 		goto err_hal_srng_deinit;
@@ -1920,6 +1923,12 @@ static void ath12k_core_restart(struct w
 	ret = ath12k_core_reconfigure_on_crash(ab);
 	if (ret) {
 		ath12k_err(ab, "failed to reconfigure driver on crash recovery\n");
+		/*
+		 * If for any reason, reconfiguration fails, issue bug on for
+		 * Mode 0
+		 */
+		if (ath12k_ssr_failsafe_mode && ath12k_recovery_mode == ATH12K_MLO_RECOVERY_MODE0)
+			BUG_ON(1);
 		return;
 	}
 
@@ -2059,6 +2068,9 @@ static void ath12k_core_reset(struct wor
 		 */
 		ath12k_warn(ab, "already resetting count %d\n", reset_count);
 
+		if (ath12k_ssr_failsafe_mode && ath12k_recovery_mode == ATH12K_MLO_RECOVERY_MODE0)
+			BUG_ON(1);
+
 		reinit_completion(&ab->reset_complete);
 		time_left = wait_for_completion_timeout(&ab->reset_complete,
 							ATH12K_RESET_TIMEOUT_HZ);
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -240,6 +240,7 @@ enum ath12k_dev_flags {
 	ATH12K_FLAG_HW_GROUP_ATTACHED,
 	ATH12K_FLAG_FTM_SEGMENTED,
 	ATH12K_FLAG_PPE_DS_ENABLED,
+	ATH12K_FLAG_WMI_INIT_DONE,
 };
 
 enum ath12k_monitor_flags {
--- a/drivers/net/wireless/ath/ath12k/htc.c
+++ b/drivers/net/wireless/ath/ath12k/htc.c
@@ -643,6 +643,8 @@ int ath12k_htc_connect_service(struct at
 
 	reinit_completion(&htc->ctl_resp);
 
+	if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags))
+		clear_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
 	status = ath12k_htc_send(htc, ATH12K_HTC_EP_0, skb);
 	if (status) {
 		kfree_skb(skb);
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -4712,7 +4712,12 @@ static void ath12k_mac_op_vif_cfg_change
 	if (changed & BSS_CHANGED_ASSOC) {
 		/* TODO Handle STA ML assoc */
 		arvif = &ahvif->deflink;
+		if (!(arvif && arvif->ar)) {
+			ath12k_err(NULL, "unable to change vif config\n");
+			return;
+		}
 		ar = arvif->ar;
+
 		if (vif->cfg.assoc)
 			ath12k_bss_assoc(ar, arvif, &vif->bss_conf);
 		else
@@ -4745,6 +4750,15 @@ static void ath12k_mac_op_bss_info_chang
 	}
 	ar = arvif->ar;
 
+	if (!ar) {
+		ath12k_info(NULL,
+			    "bss info parameter changes %llx cached to apply after vdev create on channel assign\n",
+			    changed);
+		ahvif->cache[link_id].bss_conf_changed |= changed;
+		mutex_unlock(&ah->conf_mutex);
+		return;
+	}
+
 	mutex_lock(&ar->conf_mutex);
 
 	ath12k_mac_bss_info_changed(ar, arvif, info, changed);
@@ -4787,6 +4801,7 @@ static int ath12k_mac_vdev_delete(struct
 	ar->num_created_vdevs--;
 	arvif->ahvif->num_vdev_created--;
 	arvif->is_created = false;
+	arvif->ar = NULL;
 
 clean_up:
 #ifdef CPTCFG_ATH12K_BONDED_DS_SUPPORT
@@ -4795,8 +4810,6 @@ clean_up:
 		ath12k_disable_ds_for_link_netdev(arvif, arvif->ndev_pvt->link_ndev);
 	}
 #endif
-
-	arvif->ar = NULL;
 	spin_lock_bh(&ar->data_lock);
 	list_del(&arvif->list);
 	spin_unlock_bh(&ar->data_lock);
@@ -5204,6 +5217,10 @@ static int ath12k_mac_op_hw_scan(struct
 			mutex_unlock(&ah->conf_mutex);
 			return -EINVAL;
 		} else if (ar != arvif->ar) {
+			if (!arvif->ar) {
+				mutex_unlock(&ah->conf_mutex);
+				return -EINVAL;
+			}
 			mutex_lock(&arvif->ar->conf_mutex);
 			if (arvif->ar->scan.vdev_id != -1)
 				arvif->ar->scan.vdev_id = -1;
@@ -5336,6 +5353,11 @@ static void ath12k_mac_op_cancel_hw_scan
 	}
 
 	ar = arvif->ar;
+	if (!ar) {
+		mutex_unlock(&ah->conf_mutex);
+		ath12k_err(NULL, "unable to select device to cancel scan\n");
+		return;
+	}
 	mutex_unlock(&ah->conf_mutex);
 
 	mutex_lock(&ar->conf_mutex);
@@ -5667,6 +5689,17 @@ static int ath12k_mac_op_set_key(struct
 		return 0;
 	}
 
+	if (!(arvif && arvif->ar)) {
+		ath12k_err(NULL, "Failed to set key.\n");
+		mutex_unlock(&ah->conf_mutex);
+		return -EINVAL;
+	}
+
+	if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &arvif->ar->ab->dev_flags)) {
+		mutex_unlock(&ah->conf_mutex);
+		return -ESHUTDOWN;
+	}
+
 	if (sta) {
 		ahsta = (struct ath12k_sta *)sta->drv_priv;
 		if (sta->mlo) {
@@ -6936,6 +6969,7 @@ static void ath12k_mac_ml_station_remove
 		ar = arvif->ar;
 
 		mutex_lock(&ar->conf_mutex);
+
 		ath12k_mac_station_post_remove(ar, arvif, arsta);
 		mutex_unlock(&ar->conf_mutex);
 
@@ -7268,6 +7302,11 @@ static int ath12k_mac_op_change_sta_link
 		}
 
 		ar = arvif->ar;
+		if (!ar) {
+			ath12k_err(NULL,
+				   "Failed to get ar to change sta links\n");
+			continue;
+		}
 
 		mutex_lock(&ar->conf_mutex);
 		ret = ath12k_mac_station_add(ar, arvif, arsta);
@@ -7568,6 +7607,10 @@ static int ath12k_mac_op_conf_tx(struct
 	mutex_unlock(&ah->conf_mutex);
 
 	ar = arvif->ar;
+	if (!ar) {
+		ath12k_err(NULL, "Failed to config tx\n");
+		return -EINVAL;
+	}
 
 	mutex_lock(&ar->conf_mutex);
 	ret = ath12k_mac_conf_tx(ar, arvif, ac, params);
@@ -9940,6 +9983,7 @@ err_vdev_del:
 	spin_unlock_bh(&ar->data_lock);
 
 err:
+	arvif->is_created = false;
 	arvif->ar = NULL;
 	return ret;
 }
@@ -12586,6 +12630,14 @@ static int ath12k_mac_op_set_rts_thresho
 		}
 
 		ar = arvif->ar;
+		if (!ar) {
+			ath12k_err(NULL,
+				   "Failed to set rts threshold for link: %d\n",
+				   link_id);
+			mutex_unlock(&ah->conf_mutex);
+			return ret;
+		}
+
 		ret = ath12k_set_vdev_param_to_all_vifs(ar, param_id, value);
 		if (ret) {
 			ath12k_warn(ar->ab, "failed to set RTS config for all vdevs of pdev %d",
@@ -13834,6 +13886,11 @@ static void ath12k_mac_op_sta_statistics
 
 	/* TODO accumulate link sta stats here? */
 
+	if (!ar) {
+		ath12k_err(NULL,
+			   "unable to determine sta statistics \n");
+		return;
+	}
 	sinfo->rx_duration = arsta->rx_duration;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_RX_DURATION);
 
--- a/drivers/net/wireless/ath/ath12k/mhi.c
+++ b/drivers/net/wireless/ath/ath12k/mhi.c
@@ -331,6 +331,7 @@ static void ath12k_mhi_op_status_cb(stru
 		break;
 	case MHI_CB_EE_RDDM:
 		if (!(test_bit(ATH12K_FLAG_UNREGISTERING, &ab->ag->dev_flags))) {
+			set_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
 			set_bit(ATH12K_FLAG_RECOVERY, &ab->dev_flags);
 			queue_work(ab->workqueue_aux, &ab->reset_work);
 		}
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -1546,6 +1546,8 @@ static void ath12k_pci_shutdown(struct p
 	struct ath12k_base *ab = pci_get_drvdata(pdev);
 
 	set_bit(ATH12K_FLAG_UNREGISTERING, &ab->ag->dev_flags);
+	if (test_bit(ATH12K_FLAG_RECOVERY, &ab->dev_flags))
+			return;
         cancel_work_sync(&ab->reset_work);
         ath12k_core_deinit(ab);
 }
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -387,8 +387,13 @@ int ath12k_ml_link_peers_delete(struct a
 
 		ar = arvif->ar;
 
+		if (!ar)
+			continue;
 		cancel_work_sync(&arsta->update_wk);
 
+		if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &ar->ab->dev_flags) ||
+				test_bit(ATH12K_FLAG_RECOVERY, &ar->ab->dev_flags))
+			continue;
 		mutex_lock(&ar->conf_mutex);
 		ath12k_dp_peer_cleanup(ar, arvif->vdev_id, arsta->addr);
 
@@ -414,6 +419,9 @@ int ath12k_ml_link_peers_delete(struct a
 
 		ar = arvif->ar;
 
+		if (test_bit(ATH12K_FLAG_CRASH_FLUSH, &ar->ab->dev_flags) ||
+				test_bit(ATH12K_FLAG_RECOVERY, &ar->ab->dev_flags))
+			continue;
 		mutex_lock(&ar->conf_mutex);
 		ret = ath12k_wait_for_peer_delete_done(ar, arvif->vdev_id, arsta->addr);
 		if (ret) {
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -5714,6 +5714,7 @@ static void ath12k_qmi_driver_event_work
 			break;
 		case ATH12K_QMI_EVENT_SERVER_EXIT:
 			set_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
+			clear_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags);
 			break;
 		case ATH12K_QMI_EVENT_REQUEST_MEM:
 			ret = ath12k_qmi_event_mem_request(qmi);
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -540,6 +540,10 @@ int ath12k_wmi_cmd_send(struct ath12k_pd
 	int ret = -EOPNOTSUPP;
 	struct ath12k_base *ab = wmi_sc->ab;
 
+	if (!(test_bit(ATH12K_FLAG_WMI_INIT_DONE, &wmi_sc->ab->dev_flags)) &&
+		cmd_id != WMI_INIT_CMDID)
+		return -ESHUTDOWN;
+
 	might_sleep();
 
 	if (ab->hw_params->credit_flow) {
@@ -13879,6 +13883,7 @@ void ath12k_wmi_detach(struct ath12k_bas
 	for (i = 0; i < ab->htc.wmi_ep_count; i++)
 		ath12k_wmi_pdev_detach(ab, i);
 
+	clear_bit(ATH12K_FLAG_WMI_INIT_DONE, &ab->dev_flags);
 	ath12k_wmi_free_dbring_caps(ab);
 }
 
